### PR TITLE
mlpack Header only initiative

### PR DESCRIPTION
**Pull request by shrit**
This is a first attempt to move mlpack to a header-only, I am not managing to push anything further to this PR, I think we can merge this one first and then open another one for the rest.
This is only to make the reviewing process easier.

### CODE REVIEW

**Review by conradsnicta — APPROVED** [review]

**Review by rcurtin — CHANGES_REQUESTED** [review]
(requesting changes just so mlpack-bot doesn't auto-approve and this gets accidentally merged before it's ready)

**Comment by shrit** [comment]
I will have to check for these compilation error, since I do not remember having any on my machine

**Comment by shrit** [comment]
@rcurtin, do you have an idea what is happening with the `macOS` builds?
The fix is simple, we only need to remove the `math::` and declare that we are using the mlpack::math namespace.
However, I do not understand why the error is happening in the first place, this will require more refactoring everywhere, sure I can do that but I can not understand the reason, let me know if you have an idea about it.

**Comment by rcurtin** [comment]
> @rcurtin, do you have an idea what is happening with the `macOS` builds? The fix is simple, we only need to remove the `math::` and declare that we are using the mlpack::math namespace. However, I do not understand why the error is happening in the first place, this will require more refactoring everywhere, sure I can do that but I can not understand the reason, let me know if you have an idea about it.

Just to write what we figured out while digging into this together today, the issue appears to be caused by various `using namespace`s hidden in some header files throughout the codebase.  I think if we fix those, the MacOS build should succeed... I hope! :smile:

**Comment by shrit** [comment]
@rcurtin You were right, all the CLI bindings are building normally now, I have removed all namespaces related to `mlpack::math`.
I suppose now it is fine, I am waiting on the tests to finish the build.
Once this is good, I am going to clear all the remaining namespaces inside the headers, there are still a few in the ann code.

**Comment by shrit** [comment]
@rcurtin This PR is ready for review, nothing to add from my side.

File: src/mlpack/core/kernels/epanechnikov_kernel_impl.hpp
Comment (zoq) [inline src/mlpack/core/kernels/epanechnikov_kernel_impl.hpp:122]:
```suggestion
inline double EpanechnikovKernel::GradientForSquaredDistance(
    const double distanceSquared) const
```

File: src/mlpack/core/kernels/epanechnikov_kernel_impl.hpp
Comment (zoq) [inline src/mlpack/core/kernels/epanechnikov_kernel_impl.hpp:130]:
```suggestion
  else if (distanceSquared > bandwidthSquared &&
             distanceSquared >= 0)
```
```suggestion
  else if (distanceSquared > bandwidthSquared &&
           distanceSquared >= 0)
```

File: src/mlpack/core/math/lin_alg_impl.hpp
Comment (zoq) [inline src/mlpack/core/math/lin_alg_impl.hpp:39]:
```suggestion
    if (std::abs(vec(i)) > 1e-12)
    {
      vec(i) = (vec(i) > 0) ? std::pow(vec(i), (double) power) :
          -std::pow(-vec(i), (double) power);
     }
```

File: src/mlpack/core/math/lin_alg_impl.hpp
Comment (zoq) [inline src/mlpack/core/math/lin_alg_impl.hpp:67]:
```suggestion
inline void WhitenUsingSVD(const arma::mat& x,
                           arma::mat& xWhitened,
                           arma::mat& whiteningMatrix)
```

File: src/mlpack/core/math/lin_alg_impl.hpp
Comment (zoq) [inline src/mlpack/core/math/lin_alg_impl.hpp:90]: Looks like this isn't necessary.

File: src/mlpack/core/math/lin_alg_impl.hpp
Comment (zoq) [inline src/mlpack/core/math/lin_alg_impl.hpp:150]:
```suggestion
inline void RemoveRows(const arma::mat& input,
                       const std::vector<size_t>& rowsToRemove,
                       arma::mat& output)
```

File: src/mlpack/core/math/lin_alg_impl.hpp
Comment (zoq) [inline src/mlpack/core/math/lin_alg_impl.hpp:245]:
```suggestion
```

File: src/mlpack/core/math/lin_alg_impl.hpp
Comment (zoq) [inline src/mlpack/core/math/lin_alg_impl.hpp:244]:
```suggestion
      (ceil((-1. + sqrt(1. + 8. * input.n_elem)) / 2.));
```

File: src/mlpack/core/math/random_basis.hpp
Comment (zoq) [inline src/mlpack/core/math/random_basis.hpp:32]:
```suggestion
//! Include the implementation file.
```

File: src/mlpack/core/data/detect_file_type.hpp
Comment (rcurtin) [inline src/mlpack/core/data/detect_file_type.hpp:57]:
```suggestion
inline FileType AutoDetect(std::fstream& stream,
                           const std::string& filename);
```
I know, it's pedantic. :)

File: src/mlpack/core/data/load_image.hpp
Comment (rcurtin) [inline src/mlpack/core/data/load_image.hpp:2]: If this file contains implementations, I think we should instead append them to `load_image_impl.hpp`, and move the declarations of the functions into `load_image.hpp`.

File: src/mlpack/core/data/save_image.hpp
Comment (rcurtin) [inline src/mlpack/core/data/save_image.hpp:2]: I think we should rename this file as `save_image_impl.hpp` since it contains implementations. :+1:

File: src/mlpack/core/kernels/pspectrum_string_kernel.hpp
Comment (rcurtin) [inline src/mlpack/core/kernels/pspectrum_string_kernel.hpp:131]:
```suggestion
  }
  
```

File: src/mlpack/core/kernels/pspectrum_string_kernel.hpp
Comment (rcurtin) [inline src/mlpack/core/kernels/pspectrum_string_kernel.hpp:77]: This function is a bit long---it might be a bit clearer to move it to `pspectrum_string_kernel_impl.hpp` instead. :+1:

File: src/mlpack/methods/det/dtree_impl.hpp
Comment (rcurtin) [inline src/mlpack/methods/det/dtree_impl.hpp:19]: Actually I think the intention here was to have this be in `mlpack::det::details`, so maybe it is worth doing this here?

```suggestion
namespace mlpack {
namespace det {
namespace details {
```

That would probably also mean you could revert some of the changes that add `mlpack::` or `mlpack::det::` below.

File: src/mlpack/methods/hmm/hmm_util_impl.hpp
Comment (rcurtin) [inline src/mlpack/methods/hmm/hmm_util_impl.hpp:85]:
```suggestion
          HMM<distribution::DiscreteDistribution>>(ar, x);
```
I think this could work since you are already in the `mlpack` namespace.  Correct me if I'm wrong...

File: src/mlpack/methods/hmm/hmm_util_impl.hpp
Comment (rcurtin) [inline src/mlpack/methods/hmm/hmm_util_impl.hpp:90]:
```suggestion
          HMM<distribution::GaussianDistribution>>(ar, x);
```
(Same as above.)

File: src/mlpack/methods/reinforcement_learning/q_networks/categorical_dqn.hpp
Comment (rcurtin) [inline src/mlpack/methods/reinforcement_learning/q_networks/categorical_dqn.hpp:46]:
```suggestion
  typename OutputLayerType = ann::EmptyLoss<>,
  typename InitType = ann::GaussianInitialization,
  typename NetworkType = ann::FFN<OutputLayerType, InitType>
```
Not 100% sure on this... if it does work, the same idea could apply to the rest of the file too. :+1:

File: src/mlpack/methods/reinforcement_learning/q_networks/dueling_dqn.hpp
Comment (rcurtin) [inline src/mlpack/methods/reinforcement_learning/q_networks/dueling_dqn.hpp:52]:
```suggestion
  typename OutputLayerType = ann::EmptyLoss<>,
  typename InitType = ann::GaussianInitialization,
  typename CompleteNetworkType = ann::FFN<OutputLayerType, InitType>,
  typename FeatureNetworkType = ann::Sequential<>,
  typename AdvantageNetworkType = ann::Sequential<>,
  typename ValueNetworkType = ann::Sequential<>
```
(Same as the last file.  Again, not 100% sure if it will work, but I *think* it should. :))

File: src/mlpack/methods/reinforcement_learning/q_networks/simple_dqn.hpp
Comment (rcurtin) [inline src/mlpack/methods/reinforcement_learning/q_networks/simple_dqn.hpp:32]:
```suggestion
  typename OutputLayerType = ann::MeanSquaredError<>,
  typename InitType = ann::GaussianInitialization,
  typename NetworkType = ann::FFN<OutputLayerType, InitType>
```
(Another file where maybe we can do the same thing.)

**Review by rcurtin — COMMENTED** [review]
Awesome work @shrit!  After these changes, is `libmlpack.so` now empty?  If so, I guess we could set about removing it from CMake in a future PR. :tada:

**Comment by shrit** [comment]
@rcurtin It is far from being empty there should be another two PR after this one at least to finish the transition.
There is a memory check, is it related to this PR? I think it is the case, but I am not able to figure that out correctly

File: src/mlpack/methods/reinforcement_learning/q_networks/categorical_dqn.hpp
Comment (shrit) [reply src/mlpack/methods/reinforcement_learning/q_networks/categorical_dqn.hpp:46]: Maybe, I do not remember changing these lines, I will test it locally before

**Comment by rcurtin** [comment]
> @rcurtin It is far from being empty there should be another two PR after this one at least to finish the transition. There is a memory check, is it related to this PR? I think it is the case, but I am not able to figure that out correctly

I took a look at the failing build.  I think it may be because we are defining `STB_IMAGE_WRITE_STATIC`, which marks all functions as static.  We probably need to either disable that functionality, and perhaps add `#define STBIWDEF inline`.

I found that by looking in `stb_image_write.h`:

```c++
// if STB_IMAGE_WRITE_STATIC causes problems, try defining STBIWDEF to 'inline' or 'static inline'
#ifndef STBIWDEF
#ifdef STB_IMAGE_WRITE_STATIC
#define STBIWDEF  static
#else
#ifdef __cplusplus
#define STBIWDEF  extern "C"
#else
#define STBIWDEF  extern
#endif
#endif
#endif
```

**Comment by rcurtin** [comment]
I opened https://github.com/nothings/stb/issues/1277 to see if upstream might be interested in providing the support that we need.

**Comment by shrit** [comment]
Yeah, I have seen that, therefore I commented everything related to `STB` to check if it is compiling, and it was passing locally.
The ideal solution is to find a header-only library that supports image loading, otherwise, it will become a nightmare to make STB header-only and in this case, we will have to ship our own modified STB with mlpack, which will add maintenance overhead

In fact, if we can find a header-only library for image loading this will simplify the code and remove a lot of `#ifdef` macros for finding STB, it will be like catch and it will update automatically with github action.

Also, I do not think the replacement will be that hard, maybe we need to rewrite 2 functions?

**Comment by conradsnicta** [comment]
> I am not sure what the right approach here is. Either we strip out STB and use something else that actually _is_ header-only, or perhaps we distribute a version of STB where all the function definitions are inlined (I don't like that), or we come up with some other clever solution (I can't think of any). I don't think there are any great solutions here.

I think a big part of the problem is that mlpack tries to do too many things and be too many things at the same time (ie. the classical [scope creep](https://en.wikipedia.org/wiki/Scope_creep) and [feature creep](https://en.wikipedia.org/wiki/Feature_creep) problems).  This considerably adds to the maintenance burden and slows down the development rate.  Furthermore, we end up with too many dependencies (such as STB) which cause issues such as this one.

mlpack core focus is machine learning, not loading images.  Loading images is an external problem that's best solved by the user, not mlpack.  We can't please everyone, and we shouldn't try to.

**Comment by rcurtin** [comment]
I think that this patch will fix the STB include issues: https://gist.github.com/fea124bffbf2b4ffc5f2742905eba09b .  @shrit do you want to try that and see what it does?  It builds at least locally.

> I think a big part of the problem is that mlpack tries to do too many things and be too many things at the same time (ie. the classical scope creep and feature creep problems). This considerably adds to the maintenance burden and slows down the development rate. Furthermore, we end up with too many dependencies (such as STB) which cause issues such as this one.
>
> mlpack core focus is machine learning, not loading images. Loading images is an external problem that's best solved by the user, not mlpack. We can't please everyone, and we shouldn't try to.

You make some very good points here, and this is why what I'm hoping in the longer-term is that we can split out from mlpack some kind of dataframe library or similar (or perhaps use xtensor instead).  One of the reasons that this is not entirely possible is that mlpack's target users (data scientists and similar) have high expectations for what's available.  They are often used to Python or R, where there are huge numbers of "glue" libraries available to make data processing tasks easier.  But in C++ the choices are much more limited, and for many users the prospect of manually writing STB code to load something into an Armadillo matrix is something they don't have time for, and they'll choose an alternative solution instead.  This has led us to try and roll our own support into the library.

What I would like to see us do as things move forward is a like what we did with ensmallen: we can split parts of mlpack that are unrelated to the core idea of 'a collection of efficient machine learning algorithms' into their own libraries.  Data loading and processing is certainly one of those things that is unrelated.

**Comment by shubham1206agra** [comment]
> Yeah, I have seen that, therefore I commented everything related to `STB` to check if it is compiling, and it was passing locally. The ideal solution is to find a header-only library that supports image loading, otherwise, it will become a nightmare to make STB header-only and in this case, we will have to ship our own modified STB with mlpack, which will add maintenance overhead
> 
> In fact, if we can find a header-only library for image loading this will simplify the code and remove a lot of `#ifdef` macros for finding STB, it will be like catch and it will update automatically with github action.
> 
> Also, I do not think the replacement will be that hard, maybe we need to rewrite 2 functions?

Which functions to rewrite?
Maybe I can do that.

**Comment by shrit** [comment]
@rcurtin :ok_hand: 
The patch works perfectly. It compiles without any issue locally, let us see if this passes on the CI
I am not going to add anything further for this PR, let us merge this one and then I will open another one for the remaining parts.
Whenever you have time one review should be fine.

**Comment by rcurtin** [comment]
> > Yeah, I have seen that, therefore I commented everything related to `STB` to check if it is compiling, and it was passing locally. The ideal solution is to find a header-only library that supports image loading, otherwise, it will become a nightmare to make STB header-only and in this case, we will have to ship our own modified STB with mlpack, which will add maintenance overhead
> > In fact, if we can find a header-only library for image loading this will simplify the code and remove a lot of `#ifdef` macros for finding STB, it will be like catch and it will update automatically with github action.
> > Also, I do not think the replacement will be that hard, maybe we need to rewrite 2 functions?
> 
> Which functions to rewrite? Maybe I can do that.

@shubham1206agra let's hold off on that one---I think that we can make STB work here.  Maybe someday we can refactor our data loading utilities into a separate library.

File: src/mlpack/core/data/image_info_impl.hpp
Comment (rcurtin) [inline src/mlpack/core/data/image_info_impl.hpp:22]:
```suggestion
  return std::vector<std::string>({"jpg", "png", "tga", "bmp", "psd", "gif",
      "hdr", "pic", "pnm", "jpeg"});
```
Oops, I didn't format this one correctly.

File: src/mlpack/core/data/image_info_impl.hpp
Comment (rcurtin) [inline src/mlpack/core/data/image_info_impl.hpp:28]:
```suggestion
  return std::vector<std::string>({"jpg", "png", "tga", "bmp", "hdr"});
```
This one fits all on one line. :+1:

File: src/mlpack/core/data/save_image.hpp
Comment (rcurtin) [reply src/mlpack/core/data/save_image.hpp:2]: This comment still applies---do you think we can handle it before merge?

File: src/mlpack/core/math/lin_alg_impl.hpp
Comment (rcurtin) [reply src/mlpack/core/math/lin_alg_impl.hpp:92]: Agreed, I think we can remove this line (I know it was not part of the purpose of your PR, but it's a cleanup we could do anyway).

File: src/mlpack/methods/det/dtree_impl.hpp
Comment (rcurtin) [inline src/mlpack/methods/det/dtree_impl.hpp:96]: I am not sure, but I think that you can revert most of the changes to this file, since these functions are now in the `mlpack::det` namespace.

File: src/mlpack/methods/reinforcement_learning/q_networks/categorical_dqn.hpp
Comment (rcurtin) [inline src/mlpack/methods/reinforcement_learning/q_networks/categorical_dqn.hpp:84]: I think in this file you can use `ann::` instead of `mlpack::ann`, since this is all in the `mlpack::rl` namespace (thus the `mlpack::` should be unnecessary).  Correct me if I'm wrong. :+1:

**Review by rcurtin — COMMENTED** [review]
Nice, great to see this all builds!  I only have a couple small cleanup comments, and then I agree, we should go ahead and merge it. :+1:

File: src/mlpack/core/data/load_image_impl.hpp
Comment (zoq) [inline src/mlpack/core/data/load_image_impl.hpp:106]:
```suggestion
    oss << "Currently it supports:";
```

Remove the extra space before the first supported file type.

File: src/mlpack/core/data/save_image.hpp
Comment (zoq) [inline src/mlpack/core/data/save_image.hpp:39]: I can't directly comment on the line itself, but let's switch to spaces instead of `,` to be consistent with the loading implementation, and that means we can also improve the output, so it's not `Currently image saving supports, jpg, ...` anymore.

File: src/mlpack/methods/reinforcement_learning/q_networks/simple_dqn.hpp
Comment (zoq) [inline src/mlpack/methods/reinforcement_learning/q_networks/simple_dqn.hpp:64]: Not sure I can see why this was necessary, it doesn't improve the readability.

Comment (shrit) [reply src/mlpack/methods/reinforcement_learning/q_networks/simple_dqn.hpp:64]: I am removing only `mlpack::`, but we have to keep the ann since we can not use namspaces in header file

File: src/mlpack/methods/det/dtree_impl.hpp
Comment (rcurtin) [inline src/mlpack/methods/det/dtree_impl.hpp:1038]: I think this needs a third closing brace (for the `details` namespace), but I'm not sure if that will fix the build.

File: src/mlpack/methods/det/dtree_impl.hpp
Comment (rcurtin) [inline src/mlpack/methods/det/dtree_impl.hpp:152]: Does `start` need to be initialized still too?  This could be the build failure issue, not sure.

File: src/mlpack/methods/det/dtree_impl.hpp
Comment (rcurtin) [inline src/mlpack/methods/det/dtree_impl.hpp:335]: Same comment as above about `start`. :+1:

**Review by rcurtin — COMMENTED** [review]
Everything looks good to me here---I think if you handle the comments in `dtree_impl.hpp` then the build will succeed, and once the build succeeds then from my side everything seems good and I think we should merge it. :+1:

File: src/mlpack/methods/det/dtree_impl.hpp
Comment (shrit) [reply src/mlpack/methods/det/dtree_impl.hpp:1038]: it is already closed at 148

**Comment by shrit** [comment]
@rcurtin All tests are passing now, the main issue is the memory check

File: src/mlpack/methods/det/dtree_impl.hpp
Comment (rcurtin) [reply src/mlpack/methods/det/dtree_impl.hpp:1040]: Oh, sorry!  I overlooked that.

File: CMake/TestForSTB.cmake
Comment (rcurtin) [inline CMake/TestForSTB.cmake:19]: Does this need to be `try_compile(CMAKE_HAS_STATIC_STB SOURCES stb/main.cpp stb/alib.cpp stb/blib.cpp ...)`?  I think otherwise this looks right!  I haven't tested it locally yet though.

File: CMake/stb/CMakeLists.txt
Comment (rcurtin) [inline CMake/stb/CMakeLists.txt:21]: I suppose that this is useful for local testing but we don't need to make it a part of the final commit---correct me if I'm wrong.  (I don't think this CMake configuration is used anywhere.)

File: CMake/stb/main.cpp
Comment (rcurtin) [inline CMake/stb/main.cpp:1]: It could be worth adding some comments describing what we are trying to do here:

```suggestion
// On some older STB versions, `STB_IMAGE_STATIC` does not correctly mark all
// functions as static.  So here we include two header files that in turn
// include `STB_IMAGE_IMPLEMENTATION` and `STB_IMAGE_STATIC` to check.  If
// linking fails, our STB version is not new enough.
```

File: CMake/stb/blib.hpp
Comment (rcurtin) [inline CMake/stb/blib.hpp:6]: I think that including `stb_image.h` and these definitions need to go in the .cpp files---if they're in the headers, then `stb_image.h` won't be included twice because of its header guard.

File: CMake/stb/CMakeLists.txt
Comment (shrit) [reply CMake/stb/CMakeLists.txt:21]: Not at all, I am sorry I did not even try it locally, I was working in an independent folder and then copied it to mlpack.
Just trying to verify with you if what I was doing make sense or not.

Comment (rcurtin) [reply CMake/stb/CMakeLists.txt:21]: No worries!  I think this is the right approach.  I'll play with it a bit more tomorrow. :+1:

**Comment by rcurtin** [comment]
@shrit I spent some time today trying to debug the CMake checks, and after my debugging I put them together into two patches you could apply to this branch with `git am`:

https://gist.github.com/rcurtin/4e54a7163533045cf4297dd91ba03383
https://gist.github.com/rcurtin/4fddb8784868b8e2f70f6aa985f8ac79

Take a look and let me know what you think. :+1:

**Comment by shrit** [comment]
@rcurtin Great. Thank you for the patch, much better what I have done before.
We should merge this one now, nothing to add here.

**Comment by rcurtin** [comment]
@mlpack-jenkins test this please

**Review by rcurtin — APPROVED** [review]
Everything seems fine to me now!  It will be great to get this merged.  It looks like the Go job is failing perhaps because gonum uploaded a failing version (I see the failure also in #3186).

**Comment by shrit** [comment]
Yeah, the same for the python build.
I will merge it tomorrow morning.

**Review by zoq — APPROVED** [review]
Looks good to me.

**Pull request merged**
